drivers/led/ht16k33.c

```diff
@@ -0,0 +1,282 @@
+	struct ht16k33_data *data = dev->driver_data;
+	u8_t cmd[2];
+	u8_t addr;
+	u8_t bit;
+
+	if (led >= HT16K33_DISP_SEGMENTS) {
+		return -EINVAL;
+	}
+
+	addr = led / HT16K33_DISP_COLS;
+	bit = led % HT16K33_DISP_COLS;
+
+	cmd[0] = HT16K33_CMD_DISP_DATA_ADDR | addr;
+	if (on) {
+		cmd[1] = data->buffer[addr] | BIT(bit);
+	} else {
+		cmd[1] = data->buffer[addr] & ~BIT(bit);
+	}
+
+	if (data->buffer[addr] == cmd[1]) {
+		return 0;
+	}
+
+	if (i2c_write(data->i2c, cmd, sizeof(cmd), config->i2c_addr)) {
+		LOG_ERR("Setting HT16K33 LED %s failed", on ? "on" : "off");
+		return -EIO;
+	}
+
+	data->buffer[addr] = cmd[1];
+
+	return 0;
+}
+
+static int ht16k33_led_on(struct device *dev, u32_t led)
+{
+	return ht16k33_led_set_state(dev, led, true);
+}
+
+static int ht16k33_led_off(struct device *dev, u32_t led)
+{
+	return ht16k33_led_set_state(dev, led, false);
+}
+
+static int ht16k33_init(struct device *dev)
+{
+	const struct ht16k33_cfg *config = dev->config->config_info;
+	struct ht16k33_data *data = dev->driver_data;
+	struct led_data *dev_data = &data->dev_data;
+	u8_t cmd[1 + HT16K33_DISP_DATA_SIZE]; /* 1 byte command + data */
+	int err;
+
+	data->i2c = device_get_binding(config->i2c_dev_name);
+	if (data->i2c == NULL) {
+		LOG_ERR("Failed to get I2C device");
+		return -EINVAL;
+	}
+
+	memset(&data->buffer, 0, sizeof(data->buffer));
+
+	/* Hardware specific limits */
+	dev_data->min_period = 0U;
+	dev_data->max_period = 2000U;
+	dev_data->min_brightness = 0U;
+	dev_data->max_brightness = 100U;
+
+	/* System oscillator on */
+	cmd[0] = HT16K33_CMD_SYSTEM_SETUP | HT16K33_OPT_S;
+	err = i2c_write(data->i2c, cmd, 1, config->i2c_addr);
+	if (err) {
+		LOG_ERR("Enabling HT16K33 system oscillator failed (err %d)",
+			err);
+		return -EIO;
+	}
+
+	/* Clear display RAM */
+	memset(cmd, 0, sizeof(cmd));
+	cmd[0] = HT16K33_CMD_DISP_DATA_ADDR;
+	err = i2c_write(data->i2c, cmd, sizeof(cmd), config->i2c_addr);
+	if (err) {
+		LOG_ERR("Clearing HT16K33 display RAM failed (err %d)", err);
+		return -EIO;
+	}
+
+	/* Full brightness */
+	cmd[0] = HT16K33_CMD_DIMMING_SET | 0x0f;
+	err = i2c_write(data->i2c, cmd, 1, config->i2c_addr);
+	if (err) {
+		LOG_ERR("Setting HT16K33 brightness failed (err %d)", err);
+		return -EIO;
+	}
+
+	/* Display on, blinking off */
+	cmd[0] = HT16K33_CMD_DISP_SETUP | HT16K33_OPT_D | HT16K33_OPT_BLINK_OFF;
+	err = i2c_write(data->i2c, cmd, 1, config->i2c_addr);
+	if (err) {
+		LOG_ERR("Enabling HT16K33 display failed (err %d)", err);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static const struct led_driver_api ht16k33_leds_api = {
+	.blink = ht16k33_led_blink,
+	.set_brightness = ht16k33_led_set_brightness,
+	.on = ht16k33_led_on,
+	.off = ht16k33_led_off,
+};
+
+#define HT16K33_DEVICE(id)						\
+	static const struct ht16k33_cfg ht16k33_##id##_cfg = {		\
+		.i2c_dev_name = DT_HOLTEK_HT16K33_##id##_BUS_NAME,	\
+		.i2c_addr     = DT_HOLTEK_HT16K33_##id##_BASE_ADDRESS,	\
+		.irq_enabled  = false,					\
```

Comment:
From where `irq_enabled` comes from?

Reply:
@Mani-Sadhasivam It's coming from the DTS.

Reply:
There is no such member in `ht16k33_cfg` struct!

Reply:
Ahh, now I see what you mean. It was added to the HT16K33 LED driver in the wrong commit. I have fixed that.